### PR TITLE
Change data flow

### LINKED ISSUE #117: Change the data flow in project
The indexer is the entity that kicks off the process of requesting the code.json from the agencies' website. 

The request for the code.json and the processing of the file happen after the indexer has set it's Elasticsearch parameters. This does not permit us to easily create an indexer based on the version found in the code.json.

Change the way this data follow is done to have the Indexer be instantiated once the code.json request process has the version of the file.

### PR DESCRIPTION
**Pull request by froi**
# Why

The current flow of data and code execution was very difficult to follow. It relied on a series of callbacks and heavy use of the async library waterfall functionality. This made adding support for newer versions of our code.json schemas difficult and cumbersome.

The code was heavily refactored without changing the output of the indexing process to accommodate future changes and to be more readable. Callbacks where changed to native promises to help with readability.

# What Changed

1. Refactored abstract_indexer to return promises - 02317fb
2. Initial commit AgencyJsonStream.js - 168cbc1
3. Initial commit of services/indexer/repo/RepoIndexStream.js - bf08069
4. Initial commit for RepoIndexer.js - ce72a42
5. Added validator and formatter servises - 6c4b504
6. Removed callback from indexExists - 29a2751
7. Removed unused require statements. - 7caf016
      - Removed fs
      - Removed path
8. Refactor - e590037
      - Removed unused require statements
      - Removed AgencyJsonStream from file.
      - Added AgencyJsonStream to required modules
      - Removed AgencyRepoIndexerStream from file.
      - Added RepoIndexStream to replace AgencyRepoIndexStream
      - Started to add code to handle promises.
9. Change promises fulfillment value. - 2545540
      - Fulfill for promises now only returns the response object for the promise.
10. Slight refactor.  - 7cd822e
      - Added _calculateOverallCompliance.
      - Changed _formatRepos to push formatted repo into stream one at a time.
      - Simplified callback chain in _transform.
      - _validateAgencyRepos now returns array of validated repos.
      - _getAgencyCodeJson now reads fallback data files for non production environments.
      - _saveFetchedCodeJson now expects parsed json data for its codeJson param.
      - Added lodash to the require statements.
11. Small refactor. - 094ef67
      - Added agency meta-data to validateRepo.
      - Removed comments.
      - Removed setTimeout.
12. Change writeReportToFile to promise. b9b52cb
      - Removed unused callback.
      - writeReportToFile now returns a promise.
13. Delete RepoIndexer.js - d41c3c0
      - File is not needed. RepoIndexer is defined elsewhere
14. Refactor to use indexer promises correctly. - 9346316
      - indexRepos now returns a promise and manages stream erros.
      - Renamed variables in indexRepos to have clearer names.
      - Removed lodash, config, and Writable from require statements.
15. Refactor RepoIndezStream to use promises. - 8af7df7
      - _indexRepo now returns a promise.
      - _indexRepo now uses indexer promises.
16. Linter fixes. - 1330275
      - Removed trailing comma.
      - Removed global require and now uses jsonfile to read fallback json data.
      - Fixed "line too long" warning.
      - Removed unused require statements
17. Change DHS codeUrl to fallback data - 8d066bf
      - Dev agency_endpoints.json should not go out to the internet to load data.
18. Added data/fetched directory to gitignore - ab54274
19. Remove unused callbacks and add trailing comma - c5f1eab

# Issues

Closes #117

### CODE REVIEW

File: services/indexer/abstract_indexer.js [hunk 1]
```diff
@@ -1,9 +1,9 @@
-const moment              = require("moment");
-const Logger              = require("../../utils/logger");
+const moment = require("moment");
```
Comment (zachary-kuhn) [inline services/indexer/abstract_indexer.js]: Thank you.

File: services/indexer/abstract_indexer.js [hunk 2]
```diff
@@ -44,67 +44,88 @@ class AbstractIndexer {
...
-      return callback(err, response);
+  indexDocument(doc) {
+    return new Promise((fulfill, reject) => {
+      this.client.index(doc, (err, response, status) => {
+        if(err) {
+          this.logger.error(err);
+          reject(err);
+        } else {
+          this.logger.debug(status);
+          fulfill({response, status});
```
Comment (zachary-kuhn) [inline services/indexer/abstract_indexer.js]: The other seems to be translations, but this is giving back the status as well. I assume this is intentional, but just thought I'd check.

File: services/indexer/repo/AgencyJsonStream.js [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+    agency.requirements.overallCompliance = this._calculateOverallCompliance(agency.requirements);
+    Reporter.reportRequirements(agency.acronym, agency.requirements);
+    
+    return Promise.resolve(resultRepos);
+  }
+
+  _calculateOverallCompliance(requirements){
+    //overallCompliance should be the average of the other requirements.
+    //TODO: align this approach with project-open-data's approach
+    let overallCompliance = 0;
```
Comment (zachary-kuhn) [inline services/indexer/repo/AgencyJsonStream.js]: I'm changing this in another PR, so we have to be careful when these two get merged.

File: services/indexer/repo/AgencyJsonStream.js [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+      overallCompliance += requirements[req];
+    }
+    overallCompliance /= _.size(requirements);
+    return overallCompliance;
+  }
+  
+  _formatRepos(agency, repos) {
+    logger.debug('Entered _formatCodeJson - Agency: ', agency.acronym);
+
+    repos.map(repo => {
```
Comment (zachary-kuhn) [inline services/indexer/repo/AgencyJsonStream.js]: We are `map`ping without a transform. It is better to use `forEach` in this case.

File: services/indexer/repo/AgencyJsonStream.js (hunk 3 above)
Comment (froi) [reply services/indexer/repo/AgencyJsonStream.js]: Yep. I'll let you know if there are any conflicts when I merge. I think it'll be better/easier if we merge your changes first?

File: services/indexer/abstract_indexer.js (hunk 1 above)
Comment (froi) [reply services/indexer/abstract_indexer.js]: Believe me, it was my pleasure.

Comment (froi) [reply services/indexer/abstract_indexer.js]: Believe me, it was my pleasure.

File: services/indexer/repo/AgencyJsonStream.js (hunk 3 above)
Comment (zachary-kuhn) [reply services/indexer/repo/AgencyJsonStream.js]: Opened the PR here at #132.

**Comment by froi** [comment]
[Change the data flow in project #Large [R]](https://trello.com/c/Cp0a7waK/372-change-the-data-flow-in-project-large-r)

File: config/dev/agency_endpoints.json [hunk 5]
```diff
@@ -69,7 +69,7 @@
   "id": 28,
   "name": "Department of Homeland Security",
   "acronym": "DHS",
-  "website": "https://www.dhs.gov",
+  "website": "https://www.dhs.gov",
```
Comment (zachary-kuhn) [inline config/dev/agency_endpoints.json]: Can probably undo this change.

File: services/indexer/repo/AgencyJsonStream.js [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+      requirements.agencyWidePolicy,
+      requirements.openSourceRequirement,
+      requirements.inventoryRequirement
+    ];
+    const counts = _.countBy(compliances, _.identity);
+
+    if (counts['0'] === 2) {
+      return 0;
+    } else {
+      // If all are 1 then 1 for all
```
Comment (zachary-kuhn) [inline services/indexer/repo/AgencyJsonStream.js]: Haha!

File: test/services/indexer/repo/agencyJsonStream.test.js [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+    const expectedRepos = codeJson.projects;
+    agencyJsonStream._validateAgencyRepos(agency[0], codeJson)
+      .then(repos => {
+        repos.should.be.deep.equal(expectedRepos);
+      })
+      .then(done, done);
+  });
+
+  it('Should return fomatted repos', function(done){
+    const expectedRepo = {
```
Comment (zachary-kuhn) [inline test/services/indexer/repo/agencyJsonStream.test.js]: Suggested enhancement for the future, we may want to look into pulling `faker` for generating some of these values.

Comment (froi) [reply test/services/indexer/repo/agencyJsonStream.test.js]: Good flag. I'll create an issue for it.
